yktrace/src/tir.rs

```diff
@@ -19,50 +19,97 @@ use std::{collections::HashMap, convert::TryFrom, env, io::Cursor};
+                    }
+
+                    if body.flags & bodyflags::TRACE_TAIL != 0 {
+                        trace_tails.push(body.def_id.clone());
+                    }
+
+                    let old = bodies.insert(body.def_id.clone(), body);
                     debug_assert!(old.is_none()); // should be no duplicates.
                 },
                 Pack::Debug(_) => (),
             }
         }
-        sir_map
+
+        assert!(trace_head.is_some());
+        assert_eq!(trace_tails.is_empty(), false);
+        let markers = SirMarkers { trace_head: trace_head.unwrap(), trace_tails };
+
+        Sir {bodies, markers}
     };
 }
 
+/// The serialised IR loaded in from disk. One of these structures is generated in the above
+/// `lazy_static` and is shared immutably for all threads.
+pub struct Sir {
+    /// Lets us map a DefId from a trace to a SIR body.
+    pub bodies: HashMap<DefId, Body>,
+    // Interesting locations that we need quick access to.
+    pub markers: SirMarkers
+}
+
+/// Contains the DefIds of interesting locations required for trace manipulation.
+pub struct SirMarkers {
+    /// The function which starts tracing and whose suffix gets trimmed off the top of traces.
+    /// There is only one such function: `yktrace::start_tracing()`.
+    ///
+    /// The compiler is made aware of this location by the `#[trace_head]` annotation.
+    pub trace_head: DefId,
+    /// Functions which stop tracing and whose prefix gets trimmed off the bottom of traces.
+    /// Although you'd expect only one such function, (i.e. `ThreadTracer::stop_tracing`), in fact
+    /// the inner implementation (of which there may be many) is the location which appears in the
+    /// trace (this happens even if `ThreadTracer::stop_tracing()` is marked `#[inline(never)]`).
+    /// This may be due to the fact that only one concrete struct implements `ThreadTracerImpl` at
+    /// the moment.
+    ///
+    /// The compiler is made aware of these locations by the `#[trace_tail]` annotation.
+    pub trace_tails: Vec<DefId>
+}
+
 /// A TIR trace is conceptually a straight-line path through the SIR with guarded speculation.
 #[derive(Debug)]
 pub struct TirTrace {
     ops: Vec<TirOp>
 }
 
 impl TirTrace {
-    /// Create a TirTrace from a SirTrace.
-    /// Returns Err if a DefId is encountered for which no SIR is available.
-    pub(crate) fn new<'s>(trace: &'s dyn SirTrace) -> Result<Self, InvalidTraceError> {
+    /// Create a TirTrace from a SirTrace, trimming remnants of the code which starts/stops the
+    /// tracer. Returns a TIR trace and the bounds the SIR trace was trimmed to, or Err if a DefId
+    /// is encountered for which no SIR is available.
+    pub fn new<'s>(trace: &'s dyn SirTrace) -> Result<Self, InvalidTraceError> {
         let mut ops = Vec::new();
-        let num_locs = trace.len();
+        let mut itr = trace.into_iter().peekable();
+        loop {
+            let loc = itr.next();
```

Comment:
Isn't this just a `for` loop in disguise?

Reply:
Yes, but the `peek()` later causes borrow check issues if you use a for loop.

Reply:
Ah, OK. So maybe:

```rust
loop {
    if let Some(loc) = itr.next() {
        ...
    } else {
        break;
    }
}
```

although that may also break the borrow checker... I'm just trying to avoid the conceptual and run-time overhead of `X.is_none(); ... ; X.unwrap()`.

Reply:
Your solution does work:

https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=77ecf18f4300a16e725651434c42825d

Whereas this is what I'd hoped to write:

https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=ccf28d6a3d8635b051d65e554600b442

Reply:
Ah, the borrow checker :)

Reply:
Have you tried `while let Some(loc) = iter.next() { ... }`?

Reply:
Indeed. Just got the same suggestion from someone on #rust IRC:
https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=d6ee07b93af9c802f93c9eea076b3ebb

Reply:
Good idea!

Reply:
Or even better, check this out:
https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=4f03faceeef0f67cf30e7446f3b9e00b

Credit: M-ou-se on IRC :)

Reply:
Does that work OK for the last element in the list?

Reply:
Ah, you are right. It's only gone around the loop twice. Shame.
```
[src/main.rs:5] i[0] = 1
[src/main.rs:6] i[1] = 2
[src/main.rs:5] i[0] = 2
[src/main.rs:6] i[1] = 3
```